glean-core/src/dispatcher/mod.rs

```diff
@@ -161,6 +162,28 @@ impl DispatchGuard {
             .expect("Failed to receive message on single-use channel");
     }
 
+    /// Block on the task queue emptying, with a timeout.
+    fn block_on_queue_timeout(&self, timeout: Duration) -> Result<(), RecvTimeoutError> {
+        let (tx, rx) = crossbeam_channel::bounded(0);
+
+        // We explicitly don't use `self.launch` here.
+        // We always put this task on the unbounded queue.
+        // The pre-init queue might be full before its flushed, in which case this would panic.
+        // Blocking on the queue can only work if it is eventually flushed anyway.
```

Comment:
musing aloud: This, I think, is the central idea. The only way to flush to this task would be to get through the entire preinit queue. (Theoretically, if the preinit wasn't full the "flush until this point" behaviour isn't quite preserved, as tasks enqueued after block_on_queue_timeout would need to be processed before it completes... but eh.)